### PR TITLE
virtio_serial: Add test case for empty line transfer

### PR DESCRIPTION
**Pull request by vivianQizhu**
id: 1651472, 1651526
Signed-off-by: Qianqian Zhu <qizhu@redhat.com>

### CODE REVIEW

File: qemu/tests/virtio_serial_empty_line.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    """
+    virtio_test = VirtioPortTest(test, env, params)
+
+    (vm, guest_worker, port) = virtio_test.get_vm_with_single_port()
+    session = vm.wait_for_login()
+
+    device_name = get_device_name_of_port(session, port.name)
+    try:
+        if params["sender"] == "guest":
+            # Send empty line('\n') from guest to host
```
Comment (yanan-fu) [inline qemu/tests/virtio_serial_empty_line.py]: Better explain:   `'echo' will automatically add a '\n' in the end of each writing`
```

Comment (vivianQizhu) [reply qemu/tests/virtio_serial_empty_line.py]: @yanan-fu Em...right, I will add the comment above L#85

File: qemu/tests/virtio_serial_empty_line.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            if received_data != b"\n":
+                test.fail("Received data is not same as the data sent,"
+                          " received %s, while expected '\n'"
+                          % received_data)
+            check_option = {"bytes_sent": "1"}
+        else:
+            # Send empty line('\n') from host to guest
+            port.open()
+            port.sock.send(b'\n')
+            guest_worker.cmd("virt.open('%s')" % port.name)
```
Comment (yanan-fu) [inline qemu/tests/virtio_serial_empty_line.py]: Do we need to open the port in guest first, then send date from host to it?
```
guest_worker.cmd("virt.open('%s')" % port.name)
port.sock.send(b'\n')
```

Comment (vivianQizhu) [reply qemu/tests/virtio_serial_empty_line.py]: @yanan-fu No, cause it only open the file to get file descriptor, that's a file instead of a port on guest, that does not matter if we open it in advance. While on host it is socket port, which is better to be established in advance.

File: qemu/tests/virtio_serial_empty_line.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+def get_port_info(session):
+    """
+    Get all the kernel debug info of virtio port
+    :return the dictionary of all port infos: {port_dev: {option: value}}
+    """
+    virtio_ports_path = "/sys/kernel/debug/virtio-ports/"
+    info_dict = {}
+    port_devs = session.cmd_output_safe(
+        "ls /sys/kernel/debug/virtio-ports/").split()
```
Comment (yanan-fu) [inline qemu/tests/virtio_serial_empty_line.py]: os.listdir(virtio_ports_path)

File: qemu/tests/virtio_serial_empty_line.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    Provide the device name of given port name, by checking its link
+    e.g. for qemu cmd: -device virtserialport,id=idsjuuh7,name=vs1
+        /dev/virtio-ports/vs1 -> /dev/vport0p1
+        port name(vs1) -> device name(vport0p1)
+    :param session: guest session
+    :param port_name: the name of the port, defined in qemu cmd line
+    :return: The device name of the port
+    """
+    virtio_port_path = "/dev/virtio-ports/"
+    port_path = "%s%s" % (virtio_port_path, port_name)
```
[the remark this answers is not in the document]
Comment (vivianQizhu) [reply qemu/tests/virtio_serial_empty_line.py]: @yanan-fu That's not a path exists on host, so the extra function provided by os.path.join() will not take effect, so they are actually no difference.

File: qemu/tests/virtio_serial_empty_line.py (hunk 3 above)
Comment (vivianQizhu) [reply qemu/tests/virtio_serial_empty_line.py]: @yanan-fu Same here, this is performed on guest instead of host, could not directly use `os`.

File: qemu/tests/virtio_serial_empty_line.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    Get all the kernel debug info of virtio port
+    :return the dictionary of all port infos: {port_dev: {option: value}}
+    """
+    virtio_ports_path = "/sys/kernel/debug/virtio-ports/"
+    info_dict = {}
+    port_devs = session.cmd_output_safe(
+        "ls /sys/kernel/debug/virtio-ports/").split()
+    for port_dev in port_devs:
+        port_infos = session.cmd_output("cat %s%s" % (
+            virtio_ports_path, port_dev)).splitlines()
```
Comment (yanan-fu) [inline qemu/tests/virtio_serial_empty_line.py]: Prefer `port_info = open(os.path.join(virtio_ports_path, port_dev)).readlines()`

Comment (vivianQizhu) [reply qemu/tests/virtio_serial_empty_line.py]: @yanan-fu No we can't get it this way, cause it is on guest, or that requires open up a python process on guest...

File: qemu/tests/virtio_serial_empty_line.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            unmatch_items.update({option: [real_value, exp_value]})
+    return unmatch_items
+
+
+def get_port_info(session):
+    """
+    Get all the kernel debug info of virtio port
+    :return the dictionary of all port infos: {port_dev: {option: value}}
+    """
+    virtio_ports_path = "/sys/kernel/debug/virtio-ports/"
```
Comment (yanan-fu) [inline qemu/tests/virtio_serial_empty_line.py]: L15     virtio_port_path = "/dev/virtio-ports/"
L46     virtio_ports_path = "/sys/kernel/debug/virtio-ports/"

Better use a more readable parameter name to distinguish them.

Comment (vivianQizhu) [reply qemu/tests/virtio_serial_empty_line.py]: @yanan-fu Yes,  does `virtio_port_dev_path = "/dev/virtio-ports/"` and `virtio_ports_debug_path = "/sys/kernel/debug/virtio-ports/"` look better?

File: qemu/tests/virtio_serial_empty_line.py (hunk 3 above)
Comment (yanan-fu) [reply qemu/tests/virtio_serial_empty_line.py]: oh~,  it is a cmd run with session, you are right.

File: qemu/tests/virtio_serial_empty_line.py (hunk 5 above)
Comment (yanan-fu) [reply qemu/tests/virtio_serial_empty_line.py]: same with above,   you are right.

File: qemu/tests/virtio_serial_empty_line.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (yanan-fu) [reply qemu/tests/virtio_serial_empty_line.py]: Agree with you,  no need to import one new module but only use once,  when there is one easy alternative.

File: qemu/tests/virtio_serial_empty_line.py (hunk 2 above)
Comment (yanan-fu) [reply qemu/tests/virtio_serial_empty_line.py]: got it, thanks for your explanation.

**Comment by vivianQizhu** [comment]
@yanan-fu Updated, please help review.
@sitoliu Would you please give a second review? Thanks.

**Comment by vivianQizhu** [comment]
@sitoliu Updated, thanks.

File: qemu/tests/virtio_serial_empty_line.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+            port.open()
+            # 'echo' automatically adds '\n' in the end of each writing
+            send_data_command = 'echo "" > /dev/%s' % device_name
+            session.cmd(send_data_command, timeout=120)
+            received_data = port.sock.recv(10)
+            if received_data != b"\n":
+                test.fail("Received data is not same as the data sent,"
+                          " received %s, while expected '\n'"
+                          % received_data)
+            check_option = {"bytes_sent": "1"}
```
[the remark this answers is not in the document]
Comment (vivianQizhu) [reply qemu/tests/virtio_serial_empty_line.py]: @yanan-fu I guess no, and I suppose it could be used by other modules in the future, by utilize the function `check_port_info(session, device_name, check_options)`, but won't be here in this script. So, that does not matter where `check_option` is defined.

[the remark this answers is not in the document]
Comment (vivianQizhu) [reply qemu/tests/virtio_serial_empty_line.py]: @yanan-fu And another point is, this parameter should be only configurable for the function, not configurable for the user, so prefer defining it here in the script.

[the remark this answers is not in the document]
Comment (yanan-fu) [reply qemu/tests/virtio_serial_empty_line.py]: @vivianQizhu  make sense, thanks !

Could you please provide the latest test result ?  Others LGTM.

**Comment by vivianQizhu** [comment]
@yanan-fu Latest test result, and also passed on py3.
 (1/2) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_serial_empty_line.guest_to_host: PASS (88.58 s)
 (2/2) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_serial_empty_line.host_to_guest: PASS (67.20 s)

**Comment by vivianQizhu** [comment]
Thanks all, merging it.
